File: src/AsmInstruction.cpp
#include "AsmInstruction.hpp"

using namespace std;

AsmInstruction::AsmInstruction(int number, string stringInstruction):number(number),stringInstruction(stringInstruction){
    statsNumInstruction();
}
AsmInstruction::~AsmInstruction(){

}

AsmInstruction::addInsertPosition(Function insertPosition){
    insertPositionList.push_back(insertPosition);
}

AsmInstruction::statsNumInstruction(){
    int count = 0;
    for (char ch : stringInstruction) {
        if (ch == '\n') {
            count++;
        }
    }
    numInstruction = count;
}

// vector<InsertPostion> readAndParseFile(const string& fileName) {
//     vector<InsertPostion> result;
//     ifstream file(fileName);
//     string line;

//     while (getline(file, line)) {
//         stringstream ss(line);
//         string functionName;
//         string temp;
        
//         // 读取函数名
//         if (getline(ss, functionName, '#')) {
//             InsertPostion data;
//             data.functionName = functionName;
            
//             int num_temp;
//             getline(ss, temp, ':');
//             stringstream numStream(temp);
//             numStream >> num_temp;
//             data.block_id = num_temp;
            
//             // 读取后续的数字
//             while (getline(ss, temp, '+')) {
//                 int num;
//                 stringstream numStream(temp);
//                 if (numStream >> num) {  // 只处理数字，忽略其他字符
//                     data.asmInsruction_id.push_back(num);
//                 }
//             }

//             result.push_back(data);
//         }
//     }

//     file.close();
//     return result;
// }
// int main() {
//     AsmInstruction test = AsmInstruction();
   
//     std::string fileName = "/home/lebron/IRattack/test/BasicBlock.txt";
//     auto parsedData = test.readAndParseFile(fileName);

//     // 输出解析结果
//     for (const auto& data : parsedData) {
//         std::cout << "Function Name: " << data.functionName << std::endl;
//         std::cout << "BasicBlockID: " << data.block_id << std::endl;

//         std::cout << "Numbers: ";
//         for (const auto& num : data.asmInsruction_id) {
//             std::cout << num << " ";
//         }
//         std::cout << std::endl;
//     }

//     return 0;
// }

File: py/custom_class/asminstruction.py
class AsmInstruction:
    def __init__(self, number, string_instruction):
        self.string_instruction = string_instruction  # 汇编指令字符串
        self.number = number  # 汇编指令的序号
        self.num_instruction = 0  # 指令数量，初始设为0
        self.num_insertion = 0  # 用其插入基本块的次数，初始设为0
        self.insert_positions = []  # 存储插入位置的列表 _Z9ncfution1i#0 基本块的名字

    def add_insert_position(self, insert_position):
        # 将新的插入位置添加到列表中，并增加插入次数
        self.insert_positions.append(insert_position)
        self.num_insertion += 1

    def stats_num_instruction(self):
        # 计算string_instruction中换行符的数量
        self.num_instruction = self.string_instruction.count('\n')


if __name__ == "__main__":
    # 使用示例
    asm = AsmInstruction(1, "mov eax, ebx \n mov ebx, eax\n")
    asm.add_insert_position("Function_name_1")
    asm.stats_num_instruction()

    print(asm.num_instruction)  # 输出指令数量
